File: today2/binary-tree-preorder-traversal.ts
/**
 * https://leetcode.com/problems/binary-tree-preorder-traversal/description/
 */
export { preorderTraversal, TreeNode };

function preorderTraversal(root: TreeNode | null): number[] {
  return [];
}

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

File: returned/python/today11/binary_tree_preorder_traversal.py
# https://leetcode.com/problems/binary-tree-preorder-traversal/
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []

        def dfs(root: Optional[TreeNode]):
            if not root:
                return
            result.append(root.val)
            dfs(root.left)
            dfs(root.right)
        
        dfs(root)
        return result